File: macos/Fliks/State/AppState.swift
import Foundation
import os
import AppKit

/// Central state object that orchestrates the full Fliks server
/// lifecycle: PostgreSQL → Node.js backend → ready.
@MainActor
final class AppState: ObservableObject {

    private let logger = Logger(subsystem: "app.fliks.macos", category: "AppState")

    @Published var serverState: ServerState = .stopped

    let config = ConfigStore()
    let postgresManager: PostgresManager
    let nodeManager = NodeManager()

    /// Periodic Postgres health check timer.
    private var healthCheckTimer: Timer?

    init() {
        self.postgresManager = PostgresManager(port: config.pgPort)

        // Clean shutdown on app termination.
        NotificationCenter.default.addObserver(
            forName: NSApplication.willTerminateNotification,
            object: nil,
            queue: .main
        ) { [weak self] _ in
            guard let self else { return }
            Task { await self.shutdown() }
        }

        // Start the server stack on launch.
        Task { await startAll() }
    }

    // MARK: - Lifecycle

    /// Full startup sequence: Postgres → Node.
    func startAll() async {
        do {
            try Paths.ensureAppSupportStructure()

            // 1. PostgreSQL
            serverState = .startingPostgres
            try await postgresManager.initialize()
            try await postgresManager.start()
            try await postgresManager.createDatabaseIfNeeded()

            // 2. Node.js backend
            serverState = .startingBackend
            let env = BackendEnvironment(port: config.port, dbPort: config.pgPort)

            await nodeManager.setOnCrash { [weak self] exitCode in
                Task { @MainActor in
                    self?.handleNodeCrash(exitCode: exitCode)
                }
            }

            try await nodeManager.start(config: env)

            serverState = .running
            startHealthChecks()

            // Open browser on first launch.
            if !config.hasCompletedFirstLaunch {
                openInBrowser()
                config.hasCompletedFirstLaunch = true
            }

            logger.info("Fliks is running at http://localhost:\(self.config.port)")

        } catch {
            logger.error("Startup failed: \(error.localizedDescription)")
            serverState = .error(error.localizedDescription)
        }
    }

    /// Clean shutdown: Node first, then Postgres.
    func shutdown() async {
        serverState = .stopping
        healthCheckTimer?.invalidate()
        healthCheckTimer = nil

        await nodeManager.stop()
        try? await postgresManager.stop()

        serverState = .stopped
    }

    /// Restart: stop everything, then start again.
    func restart() async {
        await shutdown()
        await startAll()
    }

    /// Open the web UI in the default browser.
    func openInBrowser() {
        let url = URL(string: "http://localhost:\(config.port)")!
        NSWorkspace.shared.open(url)
    }

    // MARK: - Crash Recovery

    private func handleNodeCrash(exitCode: Int32) {
        logger.warning("Node.js crashed with exit code \(exitCode)")
        serverState = .error("Backend crashed (exit \(exitCode))")

        // Auto-restart after a delay.
        Task {
            try? await Task.sleep(for: .seconds(3))
            guard case .error = self.serverState else { return }
            logger.info("Auto-restarting backend...")
            self.serverState = .startingBackend
            let env = BackendEnvironment(port: self.config.port, dbPort: self.config.pgPort)
            do {
                try await self.nodeManager.start(config: env)
                self.serverState = .running
            } catch {
                self.serverState = .error(error.localizedDescription)
            }
        }
    }

    // MARK: - Health Checks

    private func startHealthChecks() {
        healthCheckTimer = Timer.scheduledTimer(withTimeInterval: 10, repeats: true) { [weak self] _ in
            guard let self else { return }
            Task {
                let pgOk = await self.postgresManager.isReady()
                if !pgOk && self.serverState == .running {
                    self.logger.warning("PostgreSQL health check failed — restarting")
                    await self.restart()
                }
            }
        }
    }
}

// MARK: - NodeManager helper

extension NodeManager {
    func setOnCrash(_ handler: @escaping @Sendable (Int32) -> Void) {
        self.onCrash = handler
    }
}

File: windows/Fliks.Tray/State/AppState.cs
using System.Diagnostics;
using Fliks.Tray.Services;
using Fliks.Tray.Utilities;

namespace Fliks.Tray.State;

/// <summary>Orchestrates the full server lifecycle: PostgreSQL → Node backend
/// → ready, with crash recovery and a periodic Postgres health check.</summary>
internal sealed class AppState
{
    public ConfigStore Config { get; } = new();
    private readonly PostgresManager _postgres;
    private readonly NodeManager _node = new();
    private System.Threading.Timer? _healthTimer;

    private ServerState _state = ServerState.Stopped;

    public ServerState State
    {
        get => _state;
        private set { _state = value; StateChanged?.Invoke(value); }
    }

    /// <summary>Raised (on a background thread) whenever the state changes; the
    /// tray marshals it onto the UI thread.</summary>
    public event Action<ServerState>? StateChanged;

    public AppState()
    {
        _postgres = new PostgresManager(Config.PgPort);
        _node.OnCrash = HandleNodeCrash;
    }

    public async Task StartAllAsync()
    {
        try
        {
            AppPaths.EnsureStructure();
            Log.Info($"startup: exe={Environment.ProcessPath} repoRoot={AppPaths.RepoRoot ?? "(installed)"}");
            Log.Info($"paths: node={AppPaths.NodeExe} main={AppPaths.BackendMainJs} pgBin={AppPaths.PgBinDir} client={AppPaths.ClientDir} data={AppPaths.AppData}");

            State = ServerState.StartingPostgres;
            Log.Info("postgres: initialize");
            await _postgres.InitializeAsync();
            Log.Info("postgres: start");
            await _postgres.StartAsync();
            Log.Info("postgres: create database");
            await _postgres.CreateDatabaseIfNeededAsync();

            State = ServerState.StartingBackend;
            Log.Info("backend: start");
            await _node.StartAsync(new BackendEnvironment(Config.Port, Config.PgPort));

            State = ServerState.Running;
            Log.Info("running");
            StartHealthChecks();

            if (!Config.HasCompletedFirstLaunch)
            {
                OpenInBrowser();
                Config.HasCompletedFirstLaunch = true;
            }
        }
        catch (Exception ex)
        {
            Log.Error($"startup failed: {ex}");
            State = ServerState.Errored(ex.Message);
        }
    }

    public async Task ShutdownAsync()
    {
        State = ServerState.Stopping;
        _healthTimer?.Dispose();
        _healthTimer = null;
        await _node.StopAsync();
        await _postgres.StopAsync();
        State = ServerState.Stopped;
    }

    public async Task RestartAsync()
    {
        await ShutdownAsync();
        await StartAllAsync();
    }

    public void OpenInBrowser()
    {
        Process.Start(new ProcessStartInfo($"http://localhost:{Config.Port}")
        {
            UseShellExecute = true,
        });
    }

    public void OpenLogsFolder()
    {
        Directory.CreateDirectory(AppPaths.LogsDir);
        Process.Start(new ProcessStartInfo(AppPaths.LogsDir) { UseShellExecute = true });
    }

    private void HandleNodeCrash(int exitCode)
    {
        State = ServerState.Errored($"Backend crashed (exit {exitCode})");
        _ = Task.Run(async () =>
        {
            await Task.Delay(3000);
            if (State.Phase != ServerPhase.Error) return;
            State = ServerState.StartingBackend;
            try
            {
                await _node.StartAsync(new BackendEnvironment(Config.Port, Config.PgPort));
                State = ServerState.Running;
            }
            catch (Exception ex)
            {
                State = ServerState.Errored(ex.Message);
            }
        });
    }

    private void StartHealthChecks()
    {
        _healthTimer = new System.Threading.Timer(async _ =>
        {
            if (State.Phase != ServerPhase.Running) return;
            if (!await _postgres.IsReadyAsync() && State.Phase == ServerPhase.Running)
                await RestartAsync();
        }, null, TimeSpan.FromSeconds(10), TimeSpan.FromSeconds(10));
    }
}
